plutus-metatheory/src/RawU.lagda.md

```diff
@@ -0,0 +1,308 @@
+
+This module provides the interface between the Haskell code and the Agda code.
+It replicates the Haskell representation of Raw. In particular, for term constants,
+we need the following extension to be able to replicate the Haskell representation
+of Universes.
+
+```
+{-# OPTIONS --type-in-type #-}
+
+module RawU where
+```
+
+## Imports
+
+```
+open import Function using (_∘_)
+open import Data.Nat using (ℕ)
+open import Data.Bool using (true;false)
+open import Data.Integer using (ℤ)
+open import Data.String using (String)
+open import Data.Bool using (Bool)
+open import Relation.Binary using (DecidableEquality)
+open import Relation.Binary.PropositionalEquality using (_≡_;refl;sym;cong;cong₂)
+open import Relation.Nullary using (yes;no;¬_)
+open import Data.Unit using (⊤;tt)
+open import Data.Product using (Σ;proj₁;proj₂) renaming (_,_ to _,,_)
+
+open import Utils using (ByteString;DATA;List;_×_;_,_)
+open import Builtin using (Builtin;equals)
+open Builtin.Builtin
+
+{-# FOREIGN GHC {-# LANGUAGE GADTs #-} #-}
+{-# FOREIGN GHC import PlutusCore #-}
+{-# FOREIGN GHC import Raw #-}
+```
+
+## A (Haskell) Universe for types
+
+The following tags use type-in-type, and map more directly to the Haskell representation
+of the default universe.
+
+Tags are indexed by the real type they represent.
+The `Esc` datatype is used in the Haskell implementation to "escape" any kind into Type.
+For constants, we only care about kind *, but we need it to match the Haskell implementation.
```

Comment:
Have you tried not matching the Haskell implementation directly? Maybe a bunch of pattern synonyms like before would work?

Reply:
I tried many things. The thing you helped me with, with `TermCon` defined in Haskell worked, but was too ugly, and we had to pass through a datatype which allowed heterogenous lists, making necessary to tag and untag every element. I think the solution here is much better. And the nonsense is restricted to this module only...